File: java-elasticemail-examples/src/main/java/com/elasticemail/examples/DoubleOptinWebhook.java
package com.elasticemail.examples;

import com.elasticemail.api.ListsApi;
import com.elasticemail.client.ApiException;
import com.elasticemail.model.EmailsPayload;
import io.javalin.Javalin;
import io.javalin.http.Context;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Alternative confirmation flow driven by Elastic Email click tracking.
 * Create a webhook (see Webhooks.java) pointing at POST /double-optin/webhook.
 * When the recipient clicks the confirm link, Elastic Email reports status=Clicked
 * with the clicked URL in "target". The contact is then added to the list.
 */
public class DoubleOptinWebhook {
    private static ListsApi listsApi;
    private static String listName;
    private static String expectedToken;

    public static void main(String[] args) {
        listsApi = new ListsApi(Ee.client());
        listName = Ee.env("ELASTICEMAIL_LIST_NAME", "Newsletter");
        expectedToken = Ee.env("ELASTICEMAIL_WEBHOOK_TOKEN", "change_me");

        int port = Integer.parseInt(Ee.env("PORT", "3000"));
        Javalin app = Javalin.create().start(port);

        // Elastic Email validates the URL with a GET when the webhook is saved.
        app.get("/double-optin/webhook", ctx -> {
            if (!tokenOk(ctx.queryParam("token"))) {
                ctx.status(401).json(Map.of("error", "Invalid token"));
                return;
            }
            ctx.json(Map.of("ok", true));
        });

        app.post("/double-optin/webhook", DoubleOptinWebhook::handle);

        System.out.println("Double opt-in webhook listening on http://localhost:" + port + "/double-optin/webhook");
    }

    private static void handle(Context ctx) {
        if (!tokenOk(ctx.queryParam("token"))) {
            ctx.status(401).json(Map.of("error", "Invalid token"));
            return;
        }

        Map<String, String> event = params(ctx);
        String status = sanitize(event.get("status"));
        String target = sanitize(event.get("target"));
        String recipient = sanitize(event.get("to"));

        if (!"Clicked".equals(status) || !target.contains("/double-optin/confirm")) {
            ctx.json(Map.of("received", true, "status", status, "message", "Event ignored"));
            return;
        }

        try {
            listsApi.listsByNameContactsPost(listName, new EmailsPayload().emails(List.of(recipient)));
            System.out.println("Subscription confirmed via click: " + recipient);
            ctx.json(Map.of("received", true, "confirmed", true, "email", recipient, "list", listName));
        } catch (ApiException e) {
            Ee.printApiError("add contact to list", e);
            ctx.status(500).json(Map.of("error", Ee.errorBody(e)));
        }
    }

    /** Query string parameters first, form fields override (Elastic Email posts form-encoded). */
    private static Map<String, String> params(Context ctx) {
        Map<String, String> merged = new HashMap<>();
        ctx.queryParamMap().forEach((k, v) -> merged.put(k, v.isEmpty() ? "" : v.get(0)));
        ctx.formParamMap().forEach((k, v) -> merged.put(k, v.isEmpty() ? "" : v.get(0)));
        return merged;
    }

    private static boolean tokenOk(String token) {
        byte[] a = (token == null ? "" : token).getBytes(StandardCharsets.UTF_8);
        byte[] b = expectedToken.getBytes(StandardCharsets.UTF_8);
        return MessageDigest.isEqual(a, b);
    }

    private static String sanitize(String value) {
        return value == null ? "" : value.replaceAll("[\\r\\n]", "");
    }
}

File: kotlin-elasticemail-examples/src/main/kotlin/com/elasticemail/examples/DoubleOptinWebhook.kt
package com.elasticemail.examples

import com.elasticemail.api.ListsApi
import com.elasticemail.client.ApiException
import com.elasticemail.model.EmailsPayload
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.serialization.jackson.jackson
import io.ktor.server.application.ApplicationCall
import io.ktor.server.application.install
import io.ktor.server.engine.embeddedServer
import io.ktor.server.netty.Netty
import io.ktor.server.plugins.contentnegotiation.ContentNegotiation
import io.ktor.server.request.contentType
import io.ktor.server.request.receiveParameters
import io.ktor.server.response.respond
import io.ktor.server.routing.get
import io.ktor.server.routing.post
import io.ktor.server.routing.routing
import java.security.MessageDigest

/**
 * Alternative confirmation flow driven by Elastic Email click tracking.
 * Create a webhook (see Webhooks.kt) pointing at /double-optin/webhook.
 * When the recipient clicks the confirm link, Elastic Email reports status=Clicked
 * with the clicked URL in "target". The contact is then added to the list.
 */
fun main(args: Array<String>) {
    val listsApi = ListsApi(Ee.client())
    val listName = Ee.env("ELASTICEMAIL_LIST_NAME", "Newsletter")
    val expectedToken = Ee.env("ELASTICEMAIL_WEBHOOK_TOKEN", "change_me")
    val port = Ee.env("PORT", "3000").toInt()

    fun tokenOk(token: String?): Boolean =
        MessageDigest.isEqual(token.orEmpty().toByteArray(Charsets.UTF_8), expectedToken.toByteArray(Charsets.UTF_8))

    suspend fun handle(call: ApplicationCall) {
        if (!tokenOk(call.request.queryParameters["token"])) {
            call.respond(HttpStatusCode.Unauthorized, mapOf("error" to "Invalid token"))
            return
        }

        val event = webhookParams(call)
        val status = sanitize(event["status"])
        val target = sanitize(event["target"])
        val recipient = sanitize(event["to"])

        if (status != "Clicked" || !target.contains("/double-optin/confirm")) {
            call.respond(mapOf("received" to true, "status" to status, "message" to "Event ignored"))
            return
        }

        try {
            listsApi.listsByNameContactsPost(listName, EmailsPayload().emails(listOf(recipient)))
            println("Subscription confirmed via click: $recipient")
            call.respond(mapOf("received" to true, "confirmed" to true, "email" to recipient, "list" to listName))
        } catch (e: ApiException) {
            Ee.printApiError("add contact to list", e)
            call.respond(HttpStatusCode.InternalServerError, mapOf("error" to Ee.errorBody(e)))
        }
    }

    embeddedServer(Netty, port = port) {
        install(ContentNegotiation) { jackson() }
        routing {
            // Elastic Email validates the URL with a GET when the webhook is saved,
            // and delivers events as GET requests with the data in the query string.
            get("/double-optin/webhook") { handle(call) }
            post("/double-optin/webhook") { handle(call) }
        }
        println("Double opt-in webhook listening on http://localhost:$port/double-optin/webhook")
    }.start(wait = true)
}

/** Query string parameters first, form fields override (Elastic Email may post form-encoded). */
private suspend fun webhookParams(call: ApplicationCall): Map<String, String> {
    val merged = mutableMapOf<String, String>()
    call.request.queryParameters.forEach { k, v -> merged[k] = v.firstOrNull().orEmpty() }
    if (call.request.contentType().match(ContentType.Application.FormUrlEncoded)) {
        call.receiveParameters().forEach { k, v -> merged[k] = v.firstOrNull().orEmpty() }
    }
    return merged
}

private fun sanitize(value: String?): String = value.orEmpty().replace(Regex("[\\r\\n]"), "")
